yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnTimelineUtils.scala

```diff
@@ -0,0 +1,720 @@
+   * Important: this reverses the input in the process
+   * @param events event sequence
+   * @return spark event sequence
+   */
+  def asSparkEvents(events: Seq[TimelineEvent]): Seq[SparkListenerEvent] = {
+    events.reverse.map { event =>
+      toSparkEvent(event)
+    }
+  }
+
+  /**
+   * Build date for display in status messages
+   * @param timestamp time in milliseconds post-Epoch
+   * @param unset string to use if timestamp == 0
+   * @return a string for messages
+   */
+  def humanDateCurrentTZ(timestamp: Long, unset: String) : String = {
+    if (timestamp == 0) {
+      unset
+    } else {
+      val dateFormatter = DateFormat.getDateTimeInstance(DateFormat.DEFAULT,
+                                                          DateFormat.LONG)
+      dateFormatter.format(timestamp)
+    }
+  }
+
+  /**
+   * Short formatted time
+   * @param timestamp time in milliseconds post-Epoch
+   * @param unset string to use if timestamp == 0
+   * @return a string for messages
+   */
+  def timeShort(timestamp: Long, unset: String) : String = {
+    if (timestamp == 0) {
+      unset
+    } else {
+      val dateFormatter = DateFormat.getTimeInstance(DateFormat.SHORT)
+      dateFormatter.format(timestamp)
+    }
+  }
+
+  /**
+   * Generate the timeline entity ID from the application and attempt ID.
+   * This is required to be unique across all entities in the timeline server
+   * @param yarnAppId yarn application ID as passed in during creation
+   * @param yarnAttemptId YARN attempt ID as passed in during creation
+   * @param sparkAppId application ID in the application start event
+   * @param attemptId attempt ID in the application start event.
+   */
+  def buildEntityId(yarnAppId: ApplicationId,
+      yarnAttemptId: Option[ApplicationAttemptId],
+      sparkAppId: Option[String],
+      attemptId: Option[String]): String = {
+    yarnAttemptId match {
+      case Some(aid) => aid.toString
+      case None => yarnAppId.toString
+    }
+  }
+
+  /**
+   * Generate the application ID for use in entity fields from the application and attempt ID.
+   * @param yarnAppId yarn application ID as passed in during creation
+   * @param sparkAppId application ID as submitted in the application start event
+   */
+  def buildApplicationIdField(yarnAppId: ApplicationId, sparkAppId: Option[String]): String = {
+    yarnAppId.toString
+  }
+
+  /**
+   * Generate an attempt ID for use in the timeline entity "other/app_id" field
+   * from the application and attempt ID.
+   *
+   * This is not guaranteed to be unique across all entities. It is
+   * only required to be unique across all attempts of an application.
+   *
+   * If the application doesn't have an attempt ID, then it is
+   * an application instance which, implicitly, is single-attempt.
+   * The value [[SINGLE_ATTEMPT]] is returned
+   * @param sparkAppId application ID as submitted in the application start event
```

Comment:
its from the time I was trying to work out what was the appropriate ids to use for entities, apps & attempts; took a while to get right, and isolating in some methods (with tests) kept things under control. Given that the spark attempt Id can't be used for entity keys, I'll cut it, and, where appropriate, the same elsewhere, except for the overall `createTimelineEntity()` method which takes everything and makes the id->entity decisions.